reflection/serverreflection.go

```diff
@@ -65,8 +65,22 @@ type serverReflectionServer struct {
 
 // Register registers the server reflection service on the given gRPC server.
 func Register(s *grpc.Server) {
-	rpb.RegisterServerReflectionServer(s, &serverReflectionServer{
-		s: s,
+	register(s, s.GetServiceInfo)
+}
+
+// RegisterWithRegistrar registers the server reflection service with the given registrar.
+// The given map describes the services that will be exposed via reflection.
+func RegisterWithRegistrar(r grpc.ServiceRegistrar, svcs map[string]grpc.ServiceInfo) {
```

Comment:
Could we do something like this instead?

```go
type GRPCServer interface {
	grpc.ServiceRegistrar
	GetServiceInfo() map[string]ServiceInfo
}

func Register(s GRPCServer) {  // or a new method if this kind of API change is unacceptable
}
```

Reply:
@dfawley, this is more or less what I proposed earlier: https://github.com/grpc/grpc-go/pull/4340#discussion_r614410318

I will push a change to this branch that makes that change.

Reply:
Done. I've pushed _two_ commits.

The first one implements what you suggest above. The downside to this is that it is not _strictly_ backwards compatible, due to the signature change for `Register`.

However, it has been done in past releases of this module, such as migrating things (particularly generated code) to require a `grpc.ClientConnInterface` or `grpc.ServiceRegistrar` _instead of_ the concrete types `*grpc.ClientConn` and `*grpc.Server`.

So maybe it's okay? What do you think?

The second commit I just pushed actually makes further changes (and changes some names, in the hopes of making the API more coherent) that _are_ strictly backwards compatible.

If you prefer the first formulation, I can rollback that last commit. Let me know.

Reply:
Of the two, I prefer the former.  The only affected users would be ones referring to the function by its type, and I see little reason to this.  Plus it keeps the API as minimal as possible.  Using an interface is mildly worrying, since it means we can't extend it in the future, but we can do type assertions (to other interfaces) or introduce a new API if we ever do need to add features.  Thanks for the PR & revisions.

Reply:
Okay, SGTM. I removed that last commit to take it back to the simpler form. Anything else you'd like to see, or do you think this one is good to go?